examples/contrib/funsor/hmm.py

```diff
@@ -0,0 +1,657 @@
+    sequences = data['test']['sequences'][..., present_notes]
+    lengths = data['test']['sequence_lengths']
+    if args.truncate:
+        lengths = lengths.clamp(max=args.truncate)
+    num_observations = float(lengths.sum())
+
+    # note that since we removed unseen notes above (to make the problem a bit easier and for
+    # numerical stability) this test loss may not be directly comparable to numbers
+    # reported on this dataset elsewhere.
+    test_loss = elbo.loss(model, guide, sequences, lengths, batch_size=sequences.shape[0], include_prior=False)
+    logging.info('test loss = {}'.format(test_loss / num_observations))
+
+    # We expect models with higher capacity to perform better,
+    # but eventually overfit to the training set.
+    capacity = sum(value.reshape(-1).size(0)
+                   for value in pyro.get_param_store().values())
+    logging.info('model_{} capacity = {} parameters'.format(args.model, capacity))
+
+
+if __name__ == '__main__':
+    assert pyro.__version__.startswith('1.4.0')
+    parser = argparse.ArgumentParser(description="MAP Baum-Welch learning Bach Chorales")
+    parser.add_argument("-m", "--model", default="1", type=str,
+                        help="one of: {}".format(", ".join(sorted(models.keys()))))
+    parser.add_argument("-n", "--num-steps", default=50, type=int)
+    parser.add_argument("-b", "--batch-size", default=8, type=int)
+    parser.add_argument("-d", "--hidden-dim", default=16, type=int)
+    parser.add_argument("-nn", "--nn-dim", default=48, type=int)
+    parser.add_argument("-nc", "--nn-channels", default=2, type=int)
+    parser.add_argument("-lr", "--learning-rate", default=0.05, type=float)
+    parser.add_argument("-t", "--truncate", type=int)
+    parser.add_argument("-p", "--print-shapes", action="store_true")
+    parser.add_argument("--seed", default=0, type=int)
+    parser.add_argument('--cuda', action='store_true')
+    parser.add_argument('--jit', action='store_true')
+    parser.add_argument('--time-compilation', action='store_true')
+    parser.add_argument('-rp', '--raftery-parameterization', action='store_true')
+    parser.add_argument('--tmc', action='store_true',
+                        help="Use Tensor Monte Carlo instead of exact enumeration "
+                             "to estimate the marginal likelihood. You probably don't want to do this, "
+                             "except to see that TMC makes Monte Carlo gradient estimation feasible "
+                             "even with very large numbers of non-reparametrized variables.")
+    parser.add_argument('--tmc-num-samples', default=10, type=int)
+    parser.add_argument('--funsor', action='store_true')
+    args = parser.parse_args()
+
+    if args.funsor:
+        import funsor
+        funsor.set_backend("torch")
+        PYRO_BACKEND = "contrib.funsor"
+    else:
+        PYRO_BACKEND = "pyro"
+
+    with pyro_backend(PYRO_BACKEND):
+        main(args)
```

Comment:
Should we hard-code funsor use here, since non-funsor use is covered by examples/hmm.py? I imagine we'll be editing these side-by-side in the future (e.g. with `vim -d`)

Reply:
It's useful for comparing results across the two backends